### DISCUSSION TITLE
How to modify source for wix convert so that it generates compilable code - conditions and strings not converted correctly

### DISCUSSION THREAD

**Discussion by robe070** [wixtoolset/issues#8675] [Questions]
How might I change this V3 source so that it has the same functionality and will convert correctly?

Or is there a way to get the converter to retain the CDATA element - its far more readable?

### V3 source
```
<Control Type="Text" Id="DbmsTypeLabel" Width="100" Height="17" X="20" Y="60" Text="!(loc.AppDlg.DbmsType)">
               <Condition Action="disable"><![CDATA[DBUT_ENABLED <> "1"]]></Condition>
               <Condition Action="hide"><![CDATA[DBUT_ENABLED <> "1" AND (NOT SHOWDISABLED)]]></Condition>
</Control>
```
### Converted source
```
<Control Type="Text" Id="DbmsTypeLabel" Width="100" Height="17" X="20" Y="60" Text="!(loc.AppDlg.DbmsType)" DisableCondition="DBUT_ENABLED &lt;&gt; &quot;1&quot;" HideCondition="DBUT_ENABLED &lt;&gt; &quot;1&quot; AND (NOT SHOWDISABLED)" />
```

**Comment by bevanweiss** [comment]
For things like migrating CDATA... it's a relatively common XML annoyance.
You'll find the source code for the Converter stuff here.
https://github.com/wixtoolset/wix/blob/main/src/wix/WixToolset.Converters/WixConverter.cs

I've had to deal with retaining CDATA escaping before, but can't remember how I did it right now... I'm pretty sure Stackoverflow had the answer.
If you create a Pull Request to fix it up in the converter I'm sure it would be appreciated :)

**Reply by robe070** [reply]
I'm not a C# programmer. And we also generate much of the WiX source code. I was using the converter on those wix files that are static.

If I was to convert it I'd need to know the syntax that will now work in WiX 4.0+.

V3.0 is this:
```
<Publish Dialog="ExitDialog" Control="Finish" Event="DoAction" Value="ExecApp">WIXUI_EXITDIALOGOPTIONALCHECKBOX = 1 AND ((NOT Installed) OR UPGRADING="1")</Publish>
```

V4.0 is this:
I've attempted various syntaxes to get it to build. The xml syntax declares that CDATA may be used wherever a string might be used. And it seems there is a nuance there that I'm unaware of.
```
<Publish Dialog="ExitDialog" Control="Finish" Event="DoAction" Value="ExecApp" Condition=<![CDATA[WIXUI_EXITDIALOGOPTIONALCHECKBOX = 1 AND ((NOT Installed) OR UPGRADING="1")]]> />
```
error WIX0104: Not a valid source file; detail: '<' is an unexpected token. The expected token is '"' or '''. 
From that I imply that either " or ' must follow the '='
```
<Publish Dialog="ExitDialog" Control="Finish" Event="DoAction" Value="ExecApp" Condition="<![CDATA[WIXUI_EXITDIALOGOPTIONALCHECKBOX = 1 AND ((NOT Installed) OR UPGRADING="1")]]>" />
```
error WIX0104: Not a valid source file; detail: '<', hexadecimal value 0x3C, is an invalid attribute character.

Can CDATA still be used as a Condition in WiX 4.0?

**Reply by bevanweiss** [reply]
Unfortunately since the Condition is now an attribute, it does not support CData.  XML Attribute tags do not allow CData sections.
https://stackoverflow.com/a/1289531/2535822
I'd have to defer to Bob/Rob as to why it went from Element to Attribute for these, although I would guess that it's because it makes the syntax validation so much simpler.  Getting attributes from an XML element is simple, enumerating Elements themselves, slightly more annoying.
